### ISSUE TITLE
SPMD on TPU Pod with Multiple Machines and Randomness/Seed 

### ISSUE THREAD

**Issue by huzama** [pytorch/xla#7274]
Hello,

I have a question regarding the behavior of seeds in distributed environments. Please refer to the script below for context:

https://github.com/pytorch/xla/blob/0025ca7662f3a04e9855f965b5061f41089029a5/docs/spmd.md?plain=1#L53-L61

My specific concern is about the behavior of seeds in such settings. If I don't explicitly set a seed, will the tensor `t` differ across machines? If so, does this imply that any randomly generated parameters will also be different on each machine, potentially leading to instability during model training?

Additionally, what happens when we shard the tensor `t`? Which machine’s state will be preserved?

I couldn't find detailed information about managing randomness in the XLA documentation aside from the mentioned section, which notes that it’s generally unnecessary for TPU v4:

https://github.com/pytorch/xla/blob/0025ca7662f3a04e9855f965b5061f41089029a5/docs/pjrt.md?plain=1#L34-L52

Any insights or solutions on this matter would be greatly appreciated. Thank you!

**Comment by JackCaoG** [comment]
Seed is an input to the HLO and technically it should be replicated across all hosts. I think you need to set it on all hosts using https://github.com/pytorch/xla/blob/master/torch_xla/core/xla_model.py#L1436

@jonb377 @alanwaketan can you guys confirm?

**Comment by huzama** [comment]
@JackCaoG Despite setting the RNG state, I still get different tensors on different machines when using `torch.randn(*args)`. Setting `torch.manual_seed(seed)` fixes that, but I wonder what else I might be missing. Is there a better way to handle multiprocess SPMD?

**Comment by JackCaoG** [comment]
AFAIK random seed is the only IR(value) that was dangling around in pytorch/xla. Every other value(or XLAData) should be associated with a torch tensor hence will be either replicated or sharded.

**Comment by jonb377** [comment]
In my experiments, I've generally used `torch.manual_seed`.

Without a seed, `torch.randn(shape)` will differ across machines but  `torch.randn(shape, device='xla')` will be the same on each host (I'm not clear why it would be the same without setting a seed on the xla device, cc @JackCaoG).

I would recommend setting a seed, especially if you're initializing the model parameters on CPU. Without setting a seed, tensors which should be replicated across hosts may have different values on different hosts, since each host will generate its own copy of the tensor using its own seed.

**Comment by huzama** [comment]
Thank you @jonb377. 

May I ask what is the good practice for dealing with multiple machines in SPMD? 

Currently, I am setting the seed for all the libraries I use:

```python
random.seed(seed)
np.random.seed(seed)
torch.manual_seed(seed)
xm.set_rng_state(seed)
GENERATOR.manual_seed(seed)
```

I am not sure which ones are necessary.

**Comment by JackCaoG** [comment]
I think you at least need 
```
torch.manual_seed(seed)
xm.set_rng_state(seed)
```
to make sure pytorch cpu and torch_xla will use the same seed.
